File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed {

    public int speed;
    public int batteryDrain;
    private int charge = 100;
    private int distanceDriven = 0;

    public NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return charge <= 0;
    }

    public int distanceDriven() {
        return distanceDriven;
    }

    public void drive() {
        if (charge > 0) {
            distanceDriven += speed;
            charge -= batteryDrain;
        }

    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {

    private int distance;

    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean tryFinishTrack(NeedForSpeed car) {
        int theoreticalDrivingDistance = car.speed * (100 / car.batteryDrain);

        return theoreticalDrivingDistance >= distance;
    }
}

File: solutions/csharp/need-for-speed/1/NeedForSpeed.cs
class RemoteControlCar(int speed, int batteryDrainRate)
{
    
    
    private int _batteryLevel = 100;
    private int _distanceTravelled;
    
    public bool BatteryDrained()
    {
        return _batteryLevel < batteryDrainRate;
    }

    public int DistanceDriven()
    {
           return _distanceTravelled;
    }

    public void Drive()
    {
        if (BatteryDrained()) return;
        
        _batteryLevel -= batteryDrainRate;
        _distanceTravelled += speed;
    }

    public static RemoteControlCar Nitro()
    {
        return  new RemoteControlCar(50, 4);
    }
}

class RaceTrack(int distance)
{

    public bool TryFinishTrack(RemoteControlCar car)
    {
        while (!car.BatteryDrained())
        {
            car.Drive();
        }
        
        return car.DistanceDriven() >= distance;
    }
}
